addons/html_editor/static/src/others/x2many_media_viewer.js

```diff
@@ -0,0 +1,99 @@
+            const attachments = this.selectedMedia[this.state.activeTab];
+            const preloadedAttachments = attachments.filter((attachment) => attachment.res_model);
+            this.selectedMedia[this.state.activeTab] = attachments.filter(
+                (attachment) => !preloadedAttachments.includes(attachment)
+            );
+            if (this.selectedMedia[this.state.activeTab].length > 0) {
+                await super.save();
+                const newAttachments = this.selectedMedia[this.state.activeTab];
+                this.props.imageSave(newAttachments);
+            }
+            if (preloadedAttachments.length) {
+                this.props.imageSave(preloadedAttachments);
+            }
+        } else {
+            this.props.videoSave(this.selectedMedia[this.state.activeTab]);
+        }
+        this.props.close();
+    }
+}
+
+export class X2ManyMediaViewer extends X2ManyField {
+    static template = "html_editor.X2ManyMediaViewer";
+
+    setup() {
+        super.setup();
+        this.dialogs = useService("dialog");
+        this.orm = useService("orm");
+        this.supportedFields = ["image_1920", "image_1024", "image_512", "image_256", "image_128"];
+    }
+
+    addMedia() {
+        this.dialogs.add(X2ManyMediaDialog, {
+            save: (el) => {}, // Simple rebound to fake its execution
+            multiImages: true,
+            noDocuments: true,
+            noIcons: true,
+            imageSave: this.onImageSave.bind(this),
+            videoSave: this.onVideoSave.bind(this),
+        });
+    }
+
+    onVideoSave(videoInfo) {
+        const url = getVideoUrl(videoInfo[0].platform, videoInfo[0].videoId, videoInfo[0].params);
+        const videoList = this.props.record.data[this.props.name];
+        videoList.addNewRecord({ position: "bottom" }).then((record) => {
+            record.update({ name: videoInfo[0].platform + " - [Video]", video_url: url.href });
+        });
+    }
+
+    async onImageSave(attachments) {
+        const attachmentIds = attachments.map((attachment) => attachment.id);
+        const attachmentRecords = await this.orm.searchRead(
+            "ir.attachment",
+            [["id", "in", attachmentIds]],
+            ["id", "datas", "name"],
+            {}
+        );
+        attachmentRecords.forEach((attachment) => {
+            const imageList = this.props.record.data[this.props.name];
+            imageList.addNewRecord({ position: "bottom" }).then((record) => {
+                const activeFields = imageList.activeFields;
+                const updateData = {};
+                for (const field in activeFields) {
+                    if (attachment.datas && this.supportedFields.includes(field)) {
+                        updateData[field] = attachment.datas;
+                        updateData["name"] = attachment.name;
+                    }
+                }
+                record.update(updateData);
+            });
+        });
```

Comment:
Note that this whole bit of code creates multiple asynchronous promises and that none of them are going to awaited by anything else. In my experience this is likely to create issues if some other parts of the code have a chance to execute in the middle and wreck the state of the component, for example destroying the component. I don't know exactly what is the use case for this code so maybe my fears don't apply in this particular case but, just saying, as it is right now `onImageSave` will return way before all the images have been updated on the record and the rest of the code will continue to run without waiting for the updates.

Reply:
This piece of code is responsible for adding image(s) on the field (type StaticList in here). `addNewRecord` simply adds the images in cache, so that if user wants this can be discarded. Actual linking of attachments takes place, when manual save is clicked on the parent record. Hope this makes sense now :sweat_smile:  Let me know if you need a better explanation on it's working.